### PR TITLE
HAPI spec v3 support

### PR DESCRIPTION
**Pull request by sapols**
This PR updates the `hapi-server` code to support v3.0 of the HAPI spec rather than 2.1. The important changes are
 - `"id"` has been replaced with `"dataset"`
 - `"time.min"` has been replaced with `"start"`
 - `"time.max"` has been replaced with `"stop"`
 - a new `/about` endpoint has been added

Backwards compatibility is supported.

### CODE REVIEW

File: src/main/scala/latis/service/hapi/AboutService.scala
Comment (sapols) [inline src/main/scala/latis/service/hapi/AboutService.scala:16]: We could additionally supply a `description`, `contactId`, and `citation` here.

**Comment by sapols** [comment]
Note that I want to bring back a bunch of tests from the `latis2` branch, too: https://github.com/lasp/hapi-server/tree/latis2/src/test/scala/lasp/hapi

**Comment by lindholc** [comment]
It sounds like it is supposed to be backwards compatible. From the [spec](https://github.com/hapi-server/data-specification/blob/master/hapi-dev/HAPI-data-access-spec-dev.md#1-significant-changes-to-specification):

> HAPI 3 servers must accept both the old and these new parameter names

File: src/main/scala/latis/service/hapi/HapiService.scala
Comment (lindholc) [inline src/main/scala/latis/service/hapi/HapiService.scala:27]: I wouldn't say this groups the endpoints *into* a landing page, because a landing page isn't composed of endpoints. The landing page and the endpoints are separate things, and this service groups them together into a single service.

File: src/main/scala/latis/service/hapi/AboutResponse.scala
Comment (lindholc) [reply src/main/scala/latis/service/hapi/AboutResponse.scala:32]: I'd mark the optional values as `Option[String]` rather than using empty strings.

Circe will encode `None` to `null`, which usually isn't what we want either, but there's an easy way to drop the `null` values:

```scala
Encoder.forProduct8(...)(...).mapJson(_.dropNullValues)
```

This will produce an encoder that accepts `Option` but doesn't keep them in the JSON if they are `None`.

File: src/main/scala/latis/service/hapi/AboutResponse.scala
Comment (sapols) [inline src/main/scala/latis/service/hapi/AboutResponse.scala:61]: Not super proud of this function `f` that I'm passing to `forProduct8` below, but I can't figure out how to keep it an anonymous function and still play nicely with `.mapJson(_.dropNullValues)`. Scala seems to lose its ability to infer correct types once I tack that onto the end.

File: src/main/scala/latis/service/hapi/AboutResponse.scala
Comment (sapols) [reply src/main/scala/latis/service/hapi/AboutResponse.scala]: I couldn't just tack `.mapJson(_.dropNullValues)` onto the end of what was already there because it wouldn't compile, but it worked great after I factored out the anonymous function into its own `val`.

File: src/main/scala/latis/service/hapi/HapiService.scala
Comment (sapols) [reply src/main/scala/latis/service/hapi/HapiService.scala:27]: Oh I see that now. Sorry, I thought that `and` was a typo the first time I read it. I hope you don't mind if I add the word "together", that makes it read more clearly to me.

File: src/main/scala/latis/service/hapi/AboutResponse.scala
Comment (lindholc) [reply src/main/scala/latis/service/hapi/AboutResponse.scala:61]: Looks like you can just add a type to the second argument to `forProduct`:

```scala
Encoder.forProduct8(
    "version",
    "status",
    "id",
    "title",
    "contact",
    "description",
    "contactID",
    "citation"
  ) { (ar: AboutResponse) => (  // <-- Here
    ar.version,
    ar.status,
    ar.id,
    ar.title,
    ar.contact,
    ar.description,
    ar.contactId,
    ar.citation
  ) }.mapJson(_.dropNullValues)
```

File: build.sbt
Comment (lindholc) [inline build.sbt:3]: I'd put this in the `settings` block below.

File: src/main/scala/latis/service/hapi/AboutResponse.scala
Comment (sapols) [reply src/main/scala/latis/service/hapi/AboutResponse.scala]: Omg that's the first thing I tried. But I never put parens around `ar: AboutResponse`, apparently that was the trick.

File: build.sbt
Comment (sapols) [reply build.sbt:3]: That's how it looks in my local copy I just hadn't pushed it. I'll push now.

Comment (sapols) [reply build.sbt]: Cool. 👍

**Comment by sapols** [comment]
Whoops, close but not there yet. I added backward compatibility support but, in doing so, messed up a couple of the error status codes. The info and data services are supposed to return a 1400 if a required parameter is missing, but at the moment they're returning more granular codes (1402, 1403, etc.) about which parameters are missing (in the order they appear in the decoding pattern matches). I'll rework that.

File: src/main/scala/latis/service/hapi/DataService.scala
Comment (lindholc) [inline src/main/scala/latis/service/hapi/DataService.scala:44]: Some thoughts on this line:

Here we're trying to state that we want the value of `dataset` if it is defined, the value of `id` if it's not, or to fail with a 1400 if `id` isn't defined either (because at least one of `dataset` and `id` must be defined).

You can think of this as having two `Option`s and needing to combine them together to get a single `Option`. Luckily, we know about something that can help us combine two of the same thing together: semigroups! Specifically, the [`SemigroupK` typeclass](https://typelevel.org/cats/typeclasses/semigroupk.html).

The distinction between `Semigroup` and `SemigroupK` for `Option` is that `Semigroup` combines the things inside the `Option` while `SemigroupK` combines the `Option`s themselves. The default way to combine two `Option`s is to take the first `Some`, which is exactly what we want here:

```
@ Option("a") <+> None 
res3: Option[String] = Some("a")

@ Option("a") <+> Option("b") 
res4: Option[String] = Some("a")

@ None <+> Option("b") 
res5: Option[String] = Some("b")
```

Compare that to the `Semigroup` instance:

```
@ Option("a") |+| None 
res1: Option[String] = Some("a")

@ Option("a") |+| Option("b") 
res2: Option[String] = Some("ab")

// Can't use None directly because we need to know the type of what the Option contains.
// It worked with SemigroupK because we don't care about the contents.
@ Option.empty[String] |+| Option("b") 
res4: Option[String] = Some("b")
```

The `<+>` operator is named `combineK` and is the same thing as the `orElse` method for `Option`s.

This gets us to here:

```scala
(_dataset <+> _id).getOrElse("").asRight
```

However, we want to fail if `(_dataset <+> _id)` is `None`, because that implies that neither were set in the query. So we want to lift our `Option` into `Either` and provide the appropriate value to put in `Left` if our `Option` is `None`. That's what the `Either.fromOption` method is for:

```scala
Either.fromOption(_dataset <+> _id, Status.`1400`)
```

So this says "give me the value of `_dataset` if it's defined, `_id` if it's not defined, and fail with a 1400 status if neither are defined."

File: src/main/scala/latis/service/hapi/DataService.scala
Comment (lindholc) [inline src/main/scala/latis/service/hapi/DataService.scala:39]: This is just a thought, not something that needs to be addressed in this PR.

This line is a great example of validating (as discussed in [Parse, don't validate](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/)). It checks that `min < max`, fails if it's not, but adds no new information if it is. That means we only know by convention that `min < max`. This check could accidentally be removed or the time values could be changed after the check and we wouldn't know until something went wrong.

If we wanted to parse instead of validate, we could introduce a special pair type `(a, b)` with the invariant that `a < b`. Anywhere we have a value of that special pair type, we also have proof that the start time is before the end time. We couldn't remove this check because it would cause a type error, and we couldn't change the times in a way that violates the invariant.

File: src/main/scala/latis/service/hapi/DataService.scala
Comment (lindholc) [inline src/main/scala/latis/service/hapi/DataService.scala:48]: Try to apply what I wrote about above to these. It's a little trickier because you have two layers of types (`Option` and `ValidatedNel`), but you can do it cleanly without binding a temporary variable.

As a hint, I'll say that we don't care why a time failed to parse, so you can safely lift the `ValidatedNel` into a type that forgets the reason.

File: src/main/scala/latis/service/hapi/DataService.scala
Comment (sapols) [reply src/main/scala/latis/service/hapi/DataService.scala]: Thank you. 🙏

File: src/main/scala/latis/service/hapi/InfoService.scala
Comment (sapols) [inline src/main/scala/latis/service/hapi/InfoService.scala:28]: This works but it feels kinda funny that I'm folding twice. I tried to combine this fold with the rest of the original code (everything in the second half of this fold) such that there'd only be one fold, but it wouldn't work.

**Comment by sapols** [comment]
The code here is ready to review again. The server implements the HAPI 3.0 spec accurately. 

Remaining TODOs:
 - Get the email in `/about` from the (default?) config
 - Bring back the unit tests from the `latis2` branch

File: src/main/scala/latis/service/hapi/InfoService.scala
Comment (lindholc) [reply src/main/scala/latis/service/hapi/InfoService.scala:28]: I think it would be cleaner to get the dataset ID as a `EitherT[F, Status, String]` (take a look at the `EitherT` companion object) and then combine that with the rest using a `flatMap` or in a for comprehension.

File: src/main/scala/latis/service/hapi/DataService.scala
Comment (dlindhol) [reply src/main/scala/latis/service/hapi/DataService.scala:39]: This would be a very handy type. Unfortunately, neither scala's `Range` nor cats' `Range` enforced this. 😞 
Do you know of a handy implementation or do we need to make our own? We might want one that works with our `Data` 
and its `Ordering`. (Later.)

File: src/main/scala/latis/service/hapi/DataService.scala
Comment (lindholc) [inline src/main/scala/latis/service/hapi/DataService.scala:65]: Now that how we get the dataset identifier has changed, I think it would be cleaner to get the dataset ID from `req`.

File: src/main/scala/latis/service/hapi/DataService.scala
Comment (lindholc) [reply src/main/scala/latis/service/hapi/DataService.scala:39]: I'm not aware of a handy implementation, but it's not hard to make our own:

```scala
sealed abstract case class Lt[A](a: A, b: A)

object Lt {
  def apply[A](a: A, b: A)(implicit ord: Ordering[A]): Option[Lt[A]] =
    if (ord.lt(a, b)) Option(new Lt(a, b) {}) else None
}
```

File: src/main/scala/latis/service/hapi/DataService.scala
Comment (dlindhol) [reply src/main/scala/latis/service/hapi/DataService.scala:46]: The 1400 is if neither are present. The 1402 is if it fails to parse.

File: src/main/scala/latis/service/hapi/DataService.scala
Comment (lindholc) [reply src/main/scala/latis/service/hapi/DataService.scala:28]: More specific error codes (like 1401 instead of 1400) are optional. There's a catch-all case below that returns a 1400. I think it would take extra logic to determine if there was a missing parameter (1400) or an extra, unknown parameter (1401), so I must have opted to not use the more specific error message in this case and just use 1400 for both.

File: src/main/scala/latis/service/hapi/DataService.scala
Comment (sapols) [reply src/main/scala/latis/service/hapi/DataService.scala]: Done.

File: src/main/scala/latis/service/hapi/DataService.scala
Comment (sapols) [reply src/main/scala/latis/service/hapi/DataService.scala:53]: Yeah 1400 is most appropriate.

File: src/main/scala/latis/service/hapi/DataService.scala
Comment (sapols) [reply src/main/scala/latis/service/hapi/DataService.scala]: Oh, agreed.

File: src/main/scala/latis/service/hapi/InfoService.scala
Comment (sapols) [reply src/main/scala/latis/service/hapi/InfoService.scala:28]: Sorry can you say more about how to do this? I looked at the companion object and I'm guessing I'd use either `.fromOptionF` or `.fromOptionM`, but it's not obvious which or how.

Comment (lindholc) [reply src/main/scala/latis/service/hapi/InfoService.scala:28]: You can replace `Either.fromOption(...)` with `EitherT.fromOption[F](...)`. The `fromOptionF` and `fromOptionM` methods are for when what we're lifting into `EitherT` is already in `F`, but it's not in our case.

Once you have that, you could rewrite to look something like this:

```scala
(for {
  dataset  <- EitherT.fromOption[F](...)
  metadata <- alg.getMetadata(...).leftMap {
    // InfoError => Status
    case ...
  }
} yield metadata).foldF(
  err => ...,
  res => ...
)
```

I didn't notice this the first time, but instead of `InternalServerError` we should use `BadRequest`.

The code as it was originally written used a `fold` followed by a `flatten`, but when I was thinking about this I found out that there's a `foldF` method that will do what we want without the `flatten`.

File: src/main/resources/reference.conf
Comment (sapols) [inline src/main/resources/reference.conf:6]: We should decide on these default values together.

File: src/main/scala/latis/service/hapi/AboutService.scala
Comment (sapols) [inline src/main/scala/latis/service/hapi/AboutService.scala:27]: Any thoughts on these config paths? I went with `latis.about.*` because every other config was already under `latis` and `about` seemed natural enough. I think it's fine.

Comment (lindholc) [reply src/main/scala/latis/service/hapi/AboutService.scala:27]: I'd go with `latis.hapi.about` so we can reserve the `latis.about` namespace for our own purposes.

File: src/main/scala/latis/service/hapi/AboutService.scala
Comment (lindholc) [inline src/main/scala/latis/service/hapi/AboutService.scala:24]: I think the exceptions would be a little more helpful with the full name of the property.

File: src/main/scala/latis/service/hapi/AboutService.scala
Comment (sapols) [reply src/main/scala/latis/service/hapi/AboutService.scala]: Now; it's implemented. All are configurable.

File: src/main/scala/latis/service/hapi/InfoService.scala
Comment (sapols) [reply src/main/scala/latis/service/hapi/InfoService.scala]: Sweet thanks, that's satisfying. I'd written something nearly identical last week but the types never worked out because I was using plain `Either` instead of `EitherT`. Happy to see it work.

File: src/main/resources/reference.conf
Comment (sapols) [reply src/main/resources/reference.conf:6]: - [ ] ...and decide if we want to put any (commented out?) values in `application.conf` to help users with the syntax. I guess either way we'll need to update the `Configuration` section of the README before we merge this.

File: src/main/scala/latis/service/hapi/InfoService.scala
Comment (lindholc) [inline src/main/scala/latis/service/hapi/InfoService.scala]: We want the 1400 to correspond to a 400 but we want the 1501s (and anything we didn't explicitly handle) to correspond to a 500.

File: src/main/scala/latis/service/hapi/AboutService.scala
Comment (sapols) [reply src/main/scala/latis/service/hapi/AboutService.scala]: Done.

File: src/main/scala/latis/service/hapi/AboutService.scala
Comment (sapols) [reply src/main/scala/latis/service/hapi/AboutService.scala]: Cool, I like that.

File: src/main/resources/reference.conf
Comment (lindholc) [reply src/main/resources/reference.conf]: After thinking about it more, maybe we just don't have defaults because there aren't really reasonable defaults.

For now, that would mean that if they aren't set every request to the `/about` endpoint would result in a 500. I'd prefer that the service just fail to start, but that's probably too much refactoring for now and could be handled in another ticket.

We should at least have an [environment variable override](https://github.com/lightbend/config#optional-system-or-env-variable-overrides), because that's probably how we'll set these in production.

File: src/main/scala/latis/service/hapi/InfoService.scala
Comment (sapols) [reply src/main/scala/latis/service/hapi/InfoService.scala]: Do you mean like this?

```scala
          err => err match {
            case Status.`1400` =>
              logger.info(err.message)
              BadRequest(HapiError(err).asJson)
            case Status.`1406` | Status.`1407` =>
              logger.info(err.message)
              NotFound(HapiError(err).asJson)
            case _ =>
              logger.info(err.message)
              InternalServerError(HapiError(err).asJson)
          }
```

Comment (lindholc) [reply src/main/scala/latis/service/hapi/InfoService.scala]: Yep.

File: src/main/resources/reference.conf
Comment (sapols) [reply src/main/resources/reference.conf]: Okay, default values have been removed. And I confirmed that we get a 500 if we hit that endpoint without setting a required param.

I added environment variable overrides to `application.conf` too. The naming convention they seemed to lay out in their documentation was `FORCED_<PARAM_NAME>` so I stuck with that. E.g. the override for `id` is `FORCED_ID`.

Note that I didn't add the overrides to `reference.conf` because they don't actually override anything when you put them there. And these will only override other values specified in `application.conf` if those values are written before these overrides (when lightbend sees multiple values for a param in a file it keeps the last one to be declared).

File: src/main/resources/application.conf
Comment (lindholc) [inline src/main/resources/application.conf:16]: I think these would be better in `reference.conf` because we're saying the default value is taken from an environment variable.

We've been deriving the environment variable names by upper-casing everything and replacing periods with underscores, like `LATIS_HAPI_ABOUT`.

File: src/main/resources/reference.conf
Comment (lindholc) [reply src/main/resources/reference.conf]: I was saying "overrides" because that's what the section with the best example of the syntax was called, but I guess that's misleading because it's only a specific application of this syntax. It's really just looking for an environment variable because it can't resolve the substitution in the configuration, and using `?` allows it to fail for us to deal with later. We don't actually have anything for it to override because the environment variable itself is the default.

Comment (sapols) [reply src/main/resources/reference.conf]: Okay, I gotcha. Makes sense. And my note about its overriding behavior in `application.conf` is moot anyway because users probably won't be touching that file; our README instructs them to create their own `latis.conf` files rather than edit `application.conf`.

File: src/main/resources/application.conf
Comment (sapols) [reply src/main/resources/application.conf]: Okay.

**Comment by sapols** [comment]
Removed "[WIP]" from the title because I consider this ready to review. I'll try to add those unit tests before we merge this but as of now I've addressed all comments.

**Review by lindholc — APPROVED** [review]

File: src/main/resources/reference.conf
Comment (dlindhol) [inline src/main/resources/reference.conf:4]: Is this because they don't automatically look for env vars by munging names like LatisProperties does or is there some other precedent issue here?

Comment (sapols) [reply src/main/resources/reference.conf:4]: @lindholc ?

Comment (sapols) [reply src/main/resources/reference.conf:4]: I'm not sure.

Comment (lindholc) [reply src/main/resources/reference.conf:4]: They don't automatically look for environment variables except for substitutions (from https://github.com/lightbend/config#features-of-hocon):

> - substitutions fall back to environment variables if they don't resolve in the config itself, so `${HOME}` would work as you expect.
> - substitutions normally cause an error if unresolved, but there is a syntax `${?a.b}` to permit them to be missing.

**Review by sapols — COMMENTED** [review]
I brought back those unit tests from the `latis2` branch and added some tests to confirm backwards compatibility with the old param names (`id`, `time.min`, `time.max`).

**Pull request merged**
